src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RPC/SerializableWorkspaceEdit.ts

```diff
@@ -0,0 +1,49 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { SerializableCreateDocument } from './SerializableCreateDocument';
+import { SerializableDeleteDocument } from './SerializableDeleteDocument';
+import { SerializableRenameDocument } from './SerializableRenameDocument';
+import { SerializableTextDocumentEdit } from './SerializableTextDocumentEdit';
+import { convertTextEditFromSerializable, SerializableTextEdit } from './SerializableTextEdit';
+
+type SerializableDocumentChange = SerializableCreateDocument | SerializableRenameDocument | SerializableDeleteDocument | SerializableTextDocumentEdit;
+
+export interface SerializableWorkspaceEdit {
+    changes?: {[key: string]: Array<SerializableTextEdit>};
+    documentChanges?: Array<SerializableDocumentChange>;
+}
+
+export function convertWorkspaceEditFromSerializable(data: SerializableWorkspaceEdit): vscode.WorkspaceEdit {
+    const workspaceEdit = new vscode.WorkspaceEdit();
+
+    if (Array.isArray(data.documentChanges)) {
+        for (const documentChange of data.documentChanges) {
+            if (documentChange.kind === 'create') {
+                workspaceEdit.createFile(vscode.Uri.parse(documentChange.uri), documentChange.options);
+            } else if (documentChange.kind === 'rename') {
+                workspaceEdit.renameFile(vscode.Uri.parse(documentChange.oldUri), vscode.Uri.parse(documentChange.newUri), documentChange.options);
+            } else if (documentChange.kind === 'delete') {
+                workspaceEdit.deleteFile(vscode.Uri.parse(documentChange.uri), documentChange.options);
+            } else {
+                const changes = documentChange.edits.map(convertTextEditFromSerializable);
```

Comment:
Curious. In what cases will this else block run?

Reply:
A document change is one of four things according to the spec: CreateFile, RenameFile, DeleteFile, and TextDocumentEdit; the first three have `.kind`, so we just surmise that it has to be a document edit otherwise.